### PR TITLE
fix: [#429] Jobs Benchmark Duplicate Key

### LINKED ISSUE #429: [Bug]: Crash with default BDN output and using Jobs.
### Describe the bug

When producing benchmarks using Jobs the benchmark-publisher collected json is undifferentiated.

BDN Output
Image

Json output by BDN:
```
 {
      "DisplayInfo": "WebColdStartBenchmarks.One_Controller: ABP(Arguments=/p:DefineConstants=ABP_WEB, IterationCount=1, LaunchCount=100, RunStrategy=ColdStart, WarmupCount=0)",
      "Namespace": "RunnableBenchmarks.Web",
      "Type": "WebColdStartBenchmarks",
      "Method": "Controllers",
      "MethodTitle": "One_Controller",
      "Parameters": "",
      "FullName": "RunnableBenchmarks.Web.WebColdStartBenchmarks.Controllers",

    ...snip...
},
{
      "DisplayInfo": "WebColdStartBenchmarks.One_Controller: Native(Arguments=/p:DefineConstants=NATIVE_WEB, IterationCount=1, LaunchCount=100, RunStrategy=ColdStart, WarmupCount=0)",
      "Namespace": "RunnableBenchmarks.Web",
      "Type": "WebColdStartBenchmarks",
      "Method": "Controllers",
      "MethodTitle": "One_Controller",
      "Parameters": "",
      "FullName": "RunnableBenchmarks.Web.WebColdStartBenchmarks.Controllers",
}
```

Json from benchmarkdotnet-results-publisher

```
"date": 1757191323223,
"benches": [
  {
    "name": "RunnableBenchmarks.Web.WebColdStartBenchmarks.Controllers",
    "value": 345820155.07,
    "unit": "ns",
    "range": "± 4531260.317826246",
    "bytesAllocated": 2503056
  },
  {
    "name": "RunnableBenchmarks.Web.WebColdStartBenchmarks.Controllers",
    "value": 254318841.17,
    "unit": "ns",
    "range": "± 4054328.949877542",
    "bytesAllocated": 1399728
  },
  {
    "name": "RunnableBenchmarks.Web.WebColdStartBenchmarks.Controllers",
    "value": 243241017.19,
    "unit": "ns",
    "range": "± 6054595.118969444",
    "bytesAllocated": 1192528
  },
```

### Expected behaviour

The app to not crash.

### Actual behaviour

This Json when consumed causes a duplicate key exception, where it then proceeds to die in a fire.

### Steps to reproduce

* Use jobs in BDN
* Export using [benchmarkdotnet-results-publisher](https://github.com/martincostello/benchmarkdotnet-results-publisher)
* Try to load benchmarks application.

### Exception(s) (if any)

```text
Duplicate Key Exception in GroupBenchmarks when trying to add 


results.Add(run.Timestamp, new(run.Commit, benchmark));
```

### .NET Version

_No response_

### Anything else?

PR with one way to improve this is inbound.

* The PR simply ensures that the application doesn't crash when processing data like this.
* It cannot fix the underlying problem which is the publisher missed some important context from the BDN results.

For now I worked around the publisher problem by using a custom JSON exporter which changes the full name to `Web.WebColdStartBenchmarks.Controllers[ABP]`

### PR DESCRIPTION
**Pull request by akrock**
## Closes #429

This pull request refactors and improves the benchmark grouping and normalization logic in the dashboard, focusing on more robust handling of duplicate benchmark names and more accurate normalization of time and memory units. It also expands and strengthens the associated unit tests to cover new edge cases and behaviors.

### Benchmark grouping improvements

* Instead of throwing a duplicate key exception, the `GroupBenchmarks` method now appends numeric suffixes (e.g., `[1]`, `[2]`) to benchmark names when multiple benchmarks with the same name and timestamp are encountered, ensuring unique keys for each group. This uses a new `SuffixRegex` helper for suffix detection and incrementation. 

### Benchmark normalization enhancements

* The `NormalizeUnits` method was refactored to always normalize time values to nanoseconds first, then scale up to the most appropriate unit (ns, µs, ms, s) based on the minimum value, with a new threshold of 700 instead of 1,000. It also normalizes memory allocations to bytes first, then scales up to KB, MB, GB, or TB as needed, with more flexible handling of mixed units and missing values.

The value of 700 helps when comparing something like 800KB vs 2.5MB which allows for 0.8 MB vs 2.5 MB instead of 800 KB vs 25000 KB. 700 felt sufficiently large, but this number can be adjusted for "feel" of renders.

### Test coverage improvements

* Several new unit tests were added to `HomeTests.cs` to verify normalization for mixed memory units, correct conversion of time ranges, proper suffixing of duplicate benchmark names, and scaling of large memory units. Existing tests were updated to expect normalized units (`"ns"` for time, `"B"` for memory) rather than nulls.

### CODE REVIEW

File: src/Dashboard/Pages/Home.razor.cs [hunk 1]
```diff
@@ -69,12 +69,46 @@ public static Dictionary<string, IList<BenchmarkItem>> GroupBenchmarks(IList<Ben
...
+                        var match = SuffixRegex().Match(key);
+                        if (match.Success)
+                        {
+                            // if it does, increment the number
+                            var baseKey = match.Groups[1].Value;
+                            var number = match.Groups[3].Success switch
+                            {
+                                true => int.Parse(match.Groups[3].Value, NumberFormatInfo.InvariantInfo) + 1,
+                                false => 1,
+                            };
```
[the remark this answers is not in the document]
Comment (akrock) [reply src/Dashboard/Pages/Home.razor.cs:97]: cleaned it up some.

File: src/Dashboard/Pages/Home.razor.cs [hunk 2]
```diff
@@ -104,12 +138,70 @@ public static void NormalizeUnits(IList<BenchmarkItem> items)
...
+                var updated = item.Result.Unit switch
+                {
+                    null or "ns" => rangeDouble,
+                    "µs" => rangeDouble * 1e3,
+                    "ms" => rangeDouble * 1e6,
+                    "s" => rangeDouble * 1e9,
+                    _ => throw new ArgumentOutOfRangeException(nameof(items), item.Result.Unit, "Unknown time unit."),
+                };
+
+                item.Result.Range = prefix + updated;
```
[the remark this answers is not in the document]
Comment (akrock) [reply src/Dashboard/Pages/Home.razor.cs:173]: made it more defensive.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Sorry for the delay in reviewing - I needed time to look at it locally to check that things still rendered as before with all my own data.
